File: backend/Dtos/TodoUpdateDto.cs
using System.ComponentModel.DataAnnotations;
using backend.Models;

namespace backend.Dtos;

public class TodoUpdateDto
{
    [Required]
    [StringLength(120)]
    public string Title { get; set; } = default!;

    public bool IsCompleted { get; set; }

    [StringLength(1000)]
    public string? Description { get; set; }

    public Priority? Priority { get; set; }
    public bool Starred { get; set; } = false;
}

File: frontend/src/app/models/todo-update.dto.ts
// frontend/src/app/models/todo-update.dto.ts

import type { TodoPriority, TodoStatus } from './todo.model';

export interface UpdateTodoDto {
  /** Patchable core fields */
  title?: string;
  description?: string | null;

  /** State machine fields */
  status?: TodoStatus;
  isCompleted?: boolean;

  /** Planning / organization */
  dueDate?: string | null;
  priority?: TodoPriority;
  starred?: boolean;
  tags?: string[];
  sortOrder?: number | null;

  /** Completion / deletion lifecycle */
  completedAt?: string | null;
  isDeleted?: boolean;
  deletedAt?: string | null;

  /**
   * For audit; in a strict setup backend overrides this
   * with server time and ignores client value.
   */
  updatedAt?: string | null;

  /** Same idea as in create */
  metadata?: Record<string, unknown>;
}

